modules/FvwmDebug/FvwmDebug.man
.\" t
.\" @(#)FvwmDebug.1	1/28/94
.TH FvwmDebug 1.20 "Jan 28 1994"
.UC
.SH NAME
FvwmDebug \- the FVWM module debugger
.SH SYNOPSIS
FvwmDebug is spawned by fvwm, so no command line invocation will work.

.SH DESCRIPTION
The FvwmDebug captures the stream of module packets coming from
fvwm, and displays them on stderr.

.SH COPYRIGHTS
The FvwmDebug program, and the concept for
interfacing this module to the Window Manager, are all original work
by Robert Nation

Copyright 1994, Robert Nation. No guarantees or warranties or anything
are provided or implied in any way whatsoever. Use this program at your
own risk. Permission to use this program for any purpose is given,
as long as the copyright is kept intact.


.SH INITIALIZATION
During initialization, the FvwmDebug requests that fvwm dump a
complete window list into its pipe. These packets are terminated with
and "end Send_WindowList" message.

.SH INVOCATION
FvwmDebug can be invoked by inserting the line 'Module FvwmDebug' in
the .fvwmrc file. This can be placed on a line by itself, if FvwmDebug
is to be spawned during fvwm's initialization, or can be bound to a
menu or mouse button or keystroke to invoke it later. Fvwm will search
directory specified in the ModulePath configuration option to attempt
to locate FvwmDebug.

.SH CONFIGURATION OPTIONS
There are no configuration options.


.SH AUTHOR
Robert Nation


ChangeLog
2026-02-23 David Flater

	Renamed package to Fvwm95 Forever and moved to GitHub.
	PACKAGE_VERSION = "Forever"

	Fixed FTBs for GCC 15.
	Updated things in configure.ac, Makefile.am, and system.fvwm95rc.in.
	Converted README to Markdown and revised as necessary.
	Merged BUGS and NEVERDO into README.md.

2020-12-10 David Flater

	fvwm95-2.0.43f-dwf8

	* Fixed multiple definitions of global variables in modules
	FvwmScroll and FvwmTaskBar.  GCC 10 changed the default behavior
	from -fcommon to -fno-common, which caused these multiple
	definitions to be rejected.

2020-03-18 David Flater

	fvwm95-2.0.43f-dwf7

	* Fixed buffer overflows in fvwm/read.c SendDataToModule which
	were triggered by a long IconPath or PixmapPath.

	* Added mini-firefox.xpm.

	* Updated system.fvwm95rc.in:
	- Added Firefox;
	- Added dosbox-x to the workaround for SDL2 fullscreen lossage;
	- Disabled Alt-arrow hotkeys (shortcuts to scroll by 1/10 of a
	page) to stop conflict with Doom (Alt = strafe).

	* Updated INSTALL.

2018-01-22 David Flater

	fvwm95-2.0.43f-dwf6

	* Updated system.fvwm95rc.in:
	- Fixed usage of killall;
	- Added the fullscreen workaround to chocolate-doom etc.;
	- Comments.

	* Updated INSTALL.

2017-04-20 David Flater

	fvwm95-2.0.43f-dwf5

	* Replaced faulty recursive build system with monolithic Automake:
        - Added script utils/makemakefileam.sh to generate Makefile.am;
	- Replaced configure.ac;
	- Purged all Imakefile and Makefile.in;
	- Adjusted install and search paths to agree with modern Linux.

	* Fixed innumerable instances of overflow, underflow, corruption,
	undefined behavior, division by zero, and broken parsing which
	were responsible for crashes and misbehavior in fvwm95 and every
	module.

	* Cleaned up compiler warnings through -Wall -Wno-parentheses on
	GCC 6.3.0.  Left the following:
	  ButtonArray.c:
	    warning: statement with no effect (BOGUS)
          etc:
  	    warning: the use of `mktemp' is dangerous, better use `mkstemp'
	    warning: implicit declaration of function 'ConsoleMessage' (BOGUS)

	* Harmonized swallow triggers in FvwmButtons, FvwmWharf,
	FvwmTaskBar, and FvwmScript (SwallowExec) to match on resource
	name (Resource), resource class (Class), or window name (Name),
	but not icon name (Icon Name).

	* Fixed fvwm95 bugs:
	- windows suddenly becoming enormous when resizing occurs near the
	  edge of the virtual desktop;
	- full-screen glitch when resizing a window that was declared with
  	  NoHandles and BorderWidth 0;
	- UseStyle option not accepting quoted identifiers;
	- false matching by UseStyle option;
	- no default WindowManagerName for Restart command (now fvwm95).

	* Fixed FvwmAudio bugs:
	- playing wrong or no sounds;
	- hanging, waiting on subprocesses to exit;
	- breaking audio_filename at whitespace.

	* Fixed FvwmConsole not using libreadline by default.

	* Fixed FvwmScript and sample scripts bugs:
	- missing default for FvwmScriptPath;
	- references to non-existent things in Buttons script;
	- infinite loop in PointerSetup script (but it is hopelessly fragile
	  and may hang on another system).

	* Fixed FvwmTaskBar bugs:
	- swallowing completely broken;
	- animation broken or too fast;
	- main event loop complexity;
	- no warning on missing *FvwmTaskBarGoodyLoadableSymbol;
	- bad geometry, location, and spacing of goodies;
	- bad calibration of x coordinates on mouse clicks on goodies;
	- area cleared different than area drawn;
	- missing default for FvwmTaskBarModuleIconPathBatStat;
	- only allowing one directory in FvwmTaskBarModuleIconPathBatStat.

	* Fixed FvwmWharf bugs:
	- black box over first button when swallowing failed;
	- wrong background colors on 32-bit arch;
	- animation too fast.

	* Corrected inaccuracies in man pages and HTML docs:
	- Documented the swallow commands for FvwmTaskBar;
	- Corrected description of the swallow commands for FvwmWharf;
	- Fixed ghost reference to FvwmInitBanner;
	- Corrected description of the name-matching behavior of the Style
  	  command to agree with what is actually done in fvwm/add_window.c
	  LookInList.

	* Made a few minor additions and enhancements:
	- Replaced the sample/default system.fvwm95rc;
	- Made FvwmTaskBar goodies work by single-click;
	- Improved error reporting by GoodyLoadable.c;
	- Added mini-xemacs.xpm, mini-chrome.xpm, mini-chromium.xpm,
	  mini-wilber16.xpm, and green.xpm;
	- Added LogoTest script.

	* Cleaned up random things:
	- Merged all needed preprocessor defines into FVWMconfig.h and made
	  all sources include it;
	- Removed version.h headers and declarations that conflicted with
	  FVWMconfig.h;
	- Merged index.html and GPL.html in with the other HTML files;
	- Merged xpmroot into utils;
	- Merged and purged duplicative icon collections:
	  - FvwmScript/Scripts/some_nice_icons:
	      moved shutdown.xpm to pixmaps/shutdown2.xpm
	      moved window.xpm to pixmaps/window2.xpm
	  - FvwmTaskBar/icons:
	      moved mini-mail.xpm to mini-icons/mini-mail2.xpm
	      moved minimail.xbm to mini-icons/minimail.xbm
	- Fixed saferealloc and made UpdateString use it (3 copies);
	- Moved random READMEs, example files, etc. from module
	  subdirectories into docs/modules;
	- Moved other random READMEs, example files, etc. into docs/random;
	- Moved the untested/unmaintained ASSound and OffiX subdirs of
	  FvwmWharf to docs/modules/FvwmWharf;
	- Deleted all other unused source files and scripts;
	- Normalized file permissions;
	- Stripped trailing whitespace.

	Not done:  see BUGS, NEVERDO.

2014-01-26 David Flater

	fvwm95-2.0.43f-dwf4

	* Fixed BatStatModule showing "charging" icon instead of "online"
	when fully charged.

	* Changed the "alert" and "empty" levels from 10% and 5% to 7% and
	2% respectively.

2014-01-25 David Flater

 	fvwm95-2.0.43f-dwf3

	* Fixed crash of FvwmTaskBar when a goody (BatStatModule) returned
	NULL from its init function to signify failure to init.

	* Updated BatStatModule to work with /sys/class/power_supply.

modules/FvwmDebug/FvwmDebug.c
#include <FVWMconfig.h>
/* This module, and the entire FvwmDebug program, and the concept for
 * interfacing this module to the Window Manager, are all original work
 * by Robert Nation
 *
 * Copyright 1994, Robert Nation. No guarantees or warantees or anything
 * are provided or implied in any way whatsoever. Use this program at your
 * own risk. Permission to use this program for any purpose is given,
 * as long as the copyright is kept intact. */


#include <stdio.h>
#include <signal.h>
#include <fcntl.h>
#include <string.h>
#include <sys/wait.h>
#include <sys/time.h>
#include <unistd.h>
#include <ctype.h>
#include <stdlib.h>
#include "../../fvwm/module.h"

#include "FvwmDebug.h"

char *MyName;
int fd_width;
int fd[2];

/*
** spawn_xtee - code to execute xtee from a running executable &
** redirect stdout & stderr to it.  Currently sends both to same xtee,
** but you rewrite this to spawn off 2 xtee's - one for each stream.
*/

pid_t spawn_xtee(void)
{
  pid_t pid;
  int PIPE[2];
  char *argarray[256];

  setvbuf(stdout,NULL,_IOLBF,0); /* line buffered */

  if (pipe(PIPE))
  {
    perror("spawn_xtee");
    fprintf(stderr, "ERROR ERRATA -- Failed to create pipe for xtee.\n");
    return 0;
  }

  argarray[0] = "xtee";
  argarray[1] = "-nostdout";
  argarray[2] = NULL;

  if (!(pid = fork())) /* child */
  {
    dup2(PIPE[0], STDIN_FILENO);
    close(PIPE[0]);
    close(PIPE[1]);
    execvp("xtee",argarray);
    exit(1); /* shouldn't get here... */
  }
  else /* parent */
  {
    if (waitpid(pid, NULL, WNOHANG) != pid)
    {
      dup2(PIPE[1], STDOUT_FILENO);
      dup2(PIPE[1], STDERR_FILENO);
      close(PIPE[0]);
      close(PIPE[1]);
    }
  }

  return pid;
} /* spawn_xtee */

/***********************************************************************
 *
 *  Procedure:
 *	main - start of module
 *
 ***********************************************************************/
int main(int argc, char **argv)
{
  char *temp, *s;

  /* Save our program  name - for error messages */
  temp = argv[0];
  s=strrchr(argv[0], '/');
  if (s != NULL)
    temp = s + 1;

  MyName = safemalloc(strlen(temp)+2);
  strcpy(MyName,"*");
  strcat(MyName, temp);

  if((argc != 6)&&(argc != 7))
    {
      fprintf(stderr,"%s Version %s should only be executed by fvwm!\n",MyName,
	      VERSION);
      exit(1);
    }

  /* Dead pipe == Fvwm died */
  signal (SIGPIPE, DeadPipe);

  fd[0] = atoi(argv[1]);
  fd[1] = atoi(argv[2]);

#if 0
  spawn_xtee();
#endif

  /* Data passed in command line */
  fprintf(stderr,"Application Window 0x%s\n",argv[4]);
  fprintf(stderr,"Application Context %s\n",argv[5]);

  fd_width = GetFdWidth();

  /* Create a list of all windows */
  /* Request a list of all windows,
   * wait for ConfigureWindow packets */
  SendInfo(fd,"Send_WindowList",0);

  Loop(fd);

  return 0;
}

/***********************************************************************
 *
 *  Procedure:
 *	Loop - wait for data to process
 *
 ***********************************************************************/
void Loop(int *fd)
{
  unsigned long header[HEADER_SIZE], *body; // 2017-03-11 DWF fixed length

  while(1)
    {
      if(ReadFvwmPacket(fd[1],header,&body) > 0)
	{
	  process_message(header[1],body);
	  free(body);
	}
    }
}


/***********************************************************************
 *
 *  Procedure:
 *	Process message - examines packet types, and takes appropriate action
 *
 ***********************************************************************/
void process_message(unsigned long type,unsigned long *body)
{
  switch(type)
    {
    case M_ADD_WINDOW:
      list_add(body);
    case M_CONFIGURE_WINDOW:
      list_configure(body);
      break;
    case M_DESTROY_WINDOW:
      list_destroy(body);
      break;
    case M_FOCUS_CHANGE:
      list_focus(body);
      break;
    case M_NEW_PAGE:
      list_new_page(body);
      break;
    case M_NEW_DESK:
      list_new_desk(body);
      break;
    case M_RAISE_WINDOW:
      list_raise(body);
      break;
    case M_LOWER_WINDOW:
      list_lower(body);
      break;
    case M_ICONIFY:
      list_iconify(body);
      break;
    case M_MAP:
      list_map(body);
      break;
    case M_ICON_LOCATION:
      list_icon_loc(body);
      break;
    case M_DEICONIFY:
      list_deiconify(body);
      break;
    case M_WINDOW_NAME:
      list_window_name(body);
      break;
    case M_ICON_NAME:
      list_icon_name(body);
      break;
    case M_RES_CLASS:
      list_class(body);
      break;
    case M_RES_NAME:
      list_res_name(body);
      break;
    case M_END_WINDOWLIST:
      list_end();
      break;
    case M_DEFAULTICON:
    case M_ICON_FILE:
    default:
      list_unknown(body);
      break;
    }
}



/***********************************************************************
 *
 *  Procedure:
 *	SIGPIPE handler - SIGPIPE means fvwm is dying
 *
 ***********************************************************************/
void DeadPipe(int nonsense)
{
  fprintf(stderr,"FvwmDebug: DeadPipe\n");
  exit(0);
}

/***********************************************************************
 *
 *  Procedure:
 *	list_add - displays packet contents to stderr
 *
 ***********************************************************************/
void list_add(unsigned long *body)
{
  fprintf(stderr,"Add Window\n");
  fprintf(stderr,"\t ID %lx\n",body[0]);
  fprintf(stderr,"\t frame ID %lx\n",body[1]);
  fprintf(stderr,"\t fvwm ptr %lx\n",body[2]);
  fprintf(stderr,"\t frame x %ld\n",(long)body[3]);
  fprintf(stderr,"\t frame y %ld\n",(long)body[4]);
  fprintf(stderr,"\t frame w %ld\n",(long)body[5]);
  fprintf(stderr,"\t frame h %ld\n",(long)body[6]);
  fprintf(stderr,"\t desk %ld\n",(long)body[7]);
  fprintf(stderr,"\t flags %lx\n",body[8]);
  fprintf(stderr,"\t title height %ld\n",(long)body[9]);
  fprintf(stderr,"\t border width %ld\n",(long)body[10]);
  fprintf(stderr,"\t window base width %ld\n",(long)body[11]);
  fprintf(stderr,"\t window base height %ld\n",(long)body[12]);
  fprintf(stderr,"\t window resize width increment %ld\n",(long)body[13]);
  fprintf(stderr,"\t window resize height increment %ld\n",(long)body[14]);
  fprintf(stderr,"\t window min width %ld\n",(long)body[15]);
  fprintf(stderr,"\t window min height %ld\n",(long)body[16]);
  fprintf(stderr,"\t window max %ld\n",(long)body[17]);
  fprintf(stderr,"\t window max %ld\n",(long)body[18]);
  fprintf(stderr,"\t icon label window %lx\n",body[19]);
  fprintf(stderr,"\t icon pixmap window %lx\n",body[20]);
  fprintf(stderr,"\t window gravity %lx\n",body[21]);
}

/***********************************************************************
 *
 *  Procedure:
 *	list_configure - displays packet contents to stderr
 *
 ***********************************************************************/
void list_configure(unsigned long *body)
{
  fprintf(stderr,"Configure Window\n");
  fprintf(stderr,"\t ID %lx\n",body[0]);
  fprintf(stderr,"\t frame ID %lx\n",body[1]);
  fprintf(stderr,"\t fvwm ptr %lx\n",body[2]);
  fprintf(stderr,"\t frame x %ld\n",(long)body[3]);
  fprintf(stderr,"\t frame y %ld\n",(long)body[4]);
  fprintf(stderr,"\t frame w %ld\n",(long)body[5]);
  fprintf(stderr,"\t frame h %ld\n",(long)body[6]);
  fprintf(stderr,"\t desk %ld\n",(long)body[7]);
  fprintf(stderr,"\t flags %lx\n",body[8]);
  fprintf(stderr,"\t title height %ld\n",(long)body[9]);
  fprintf(stderr,"\t border width %ld\n",(long)body[10]);
  fprintf(stderr,"\t window base width %ld\n",(long)body[11]);
  fprintf(stderr,"\t window base height %ld\n",(long)body[12]);
  fprintf(stderr,"\t window resize width increment %ld\n",(long)body[13]);
  fprintf(stderr,"\t window resize height increment %ld\n",(long)body[14]);
  fprintf(stderr,"\t window min width %ld\n",(long)body[15]);
  fprintf(stderr,"\t window min height %ld\n",(long)body[16]);
  fprintf(stderr,"\t window max %ld\n",(long)body[17]);
  fprintf(stderr,"\t window max %ld\n",(long)body[18]);
  fprintf(stderr,"\t icon label window %lx\n",body[19]);
  fprintf(stderr,"\t icon pixmap window %lx\n",body[20]);
  fprintf(stderr,"\t window gravity %lx\n",body[21]);
}

/***********************************************************************
 *
 *  Procedure:
 *	list_destroy - displays packet contents to stderr
 *
 ***********************************************************************/
void list_destroy(unsigned long *body)
{
  fprintf(stderr,"destroy\n");
  fprintf(stderr,"\t ID %lx\n",body[0]);
  fprintf(stderr,"\t frame ID %lx\n",body[1]);
  fprintf(stderr,"\t fvwm ptr %lx\n",body[2]);
}

/***********************************************************************
 *
 *  Procedure:
 *	list_focus - displays packet contents to stderr
 *
 ***********************************************************************/
void list_focus(unsigned long *body)
{
  fprintf(stderr,"focus\n");
  fprintf(stderr,"\t ID %lx\n",body[0]);
  fprintf(stderr,"\t frame ID %lx\n",body[1]);
  fprintf(stderr,"\t fvwm ptr %lx\n",body[2]);

}

/***********************************************************************
 *
 *  Procedure:
 *	list_new_page - displays packet contents to stderr
 *
 ***********************************************************************/
void list_new_page(unsigned long *body)
{
  fprintf(stderr,"new page\n");
  fprintf(stderr,"\t x %ld\n",(long)body[0]);
  fprintf(stderr,"\t y %ld\n",(long)body[1]);
  fprintf(stderr,"\t desk %ld\n",(long)body[2]);
}

/***********************************************************************
 *
 *  Procedure:
 *	list_new_desk - displays packet contents to stderr
 *
 ***********************************************************************/
void list_new_desk(unsigned long *body)
{
  fprintf(stderr,"new desk\n");
  fprintf(stderr,"\t desk %ld\n",(long)body[0]);
}

/***********************************************************************
 *
 *  Procedure:
 *	list_raise - displays packet contents to stderr
 *
 ***********************************************************************/
void list_raise(unsigned long *body)
{
  fprintf(stderr,"raise\n");
  fprintf(stderr,"\t ID %lx\n",body[0]);
  fprintf(stderr,"\t frame ID %lx\n",body[1]);
  fprintf(stderr,"\t fvwm ptr %lx\n",body[2]);
}


/***********************************************************************
 *
 *  Procedure:
 *	list_lower - displays packet contents to stderr
 *
 ***********************************************************************/
void list_lower(unsigned long *body)
{
  fprintf(stderr,"lower\n");
  fprintf(stderr,"\t ID %lx\n",body[0]);
  fprintf(stderr,"\t frame ID %lx\n",body[1]);
  fprintf(stderr,"\t fvwm ptr %lx\n",body[2]);
}


/***********************************************************************
 *
 *  Procedure:
 *	list_unknow - handles an unrecognized packet.
 *
 ***********************************************************************/
void list_unknown(unsigned long *body)
{
  fprintf(stderr,"Unknown packet type\n");
}

/***********************************************************************
 *
 *  Procedure:
 *	list_iconify - displays packet contents to stderr
 *
 ***********************************************************************/
void list_iconify(unsigned long *body)
{
  fprintf(stderr,"iconify\n");
  fprintf(stderr,"\t ID %lx\n",body[0]);
  fprintf(stderr,"\t frame ID %lx\n",body[1]);
  fprintf(stderr,"\t fvwm ptr %lx\n",body[2]);
  fprintf(stderr,"\t icon x %ld\n",(long)body[3]);
  fprintf(stderr,"\t icon y %ld\n",(long)body[4]);
  fprintf(stderr,"\t icon w %ld\n",(long)body[5]);
  fprintf(stderr,"\t icon h %ld\n",(long)body[6]);
}


/***********************************************************************
 *
 *  Procedure:
 *	list_map - displays packet contents to stderr
 *
 ***********************************************************************/
void list_map(unsigned long *body)
{
  fprintf(stderr,"map\n");
  fprintf(stderr,"\t ID %lx\n",body[0]);
  fprintf(stderr,"\t frame ID %lx\n",body[1]);
  fprintf(stderr,"\t fvwm ptr %lx\n",body[2]);
}


/***********************************************************************
 *
 *  Procedure:
 *	list_icon_loc - displays packet contents to stderr
 *
 ***********************************************************************/
void list_icon_loc(unsigned long *body)
{
  fprintf(stderr,"icon location\n");
  fprintf(stderr,"\t ID %lx\n",body[0]);
  fprintf(stderr,"\t frame ID %lx\n",body[1]);
  fprintf(stderr,"\t fvwm ptr %lx\n",body[2]);
  fprintf(stderr,"\t icon x %ld\n",(long)body[3]);
  fprintf(stderr,"\t icon y %ld\n",(long)body[4]);
  fprintf(stderr,"\t icon w %ld\n",(long)body[5]);
  fprintf(stderr,"\t icon h %ld\n",(long)body[6]);
}



/***********************************************************************
 *
 *  Procedure:
 *	list_deiconify - displays packet contents to stderr
 *
 ***********************************************************************/

void list_deiconify(unsigned long *body)
{
  fprintf(stderr,"de-iconify\n");
  fprintf(stderr,"\t ID %lx\n",body[0]);
  fprintf(stderr,"\t frame ID %lx\n",body[1]);
  fprintf(stderr,"\t fvwm ptr %lx\n",body[2]);
}

/***********************************************************************
 *
 *  Procedure:
 *	list_window_name - displays packet contents to stderr
 *
 ***********************************************************************/

void list_window_name(unsigned long *body)
{
  fprintf(stderr,"window name\n");
  fprintf(stderr,"\t ID %lx\n",body[0]);
  fprintf(stderr,"\t frame ID %lx\n",body[1]);
  fprintf(stderr,"\t fvwm ptr %lx\n",body[2]);
  fprintf(stderr,"\t window name %s\n",(char *)(&body[3]));

}


/***********************************************************************
 *
 *  Procedure:
 *	list_icon_name - displays packet contents to stderr
 *
 ***********************************************************************/
void list_icon_name(unsigned long *body)
{
  fprintf(stderr,"icon name\n");
  fprintf(stderr,"\t ID %lx\n",body[0]);
  fprintf(stderr,"\t frame ID %lx\n",body[1]);
  fprintf(stderr,"\t fvwm ptr %lx\n",body[2]);
  fprintf(stderr,"\t icon name %s\n",(char *)(&body[3]));
}



/***********************************************************************
 *
 *  Procedure:
 *	list_class - displays packet contents to stderr
 *
 ***********************************************************************/
void list_class(unsigned long *body)
{
  fprintf(stderr,"window class\n");
  fprintf(stderr,"\t ID %lx\n",body[0]);
  fprintf(stderr,"\t frame ID %lx\n",body[1]);
  fprintf(stderr,"\t fvwm ptr %lx\n",body[2]);
  fprintf(stderr,"\t window class %s\n",(char *)(&body[3]));
}


/***********************************************************************
 *
 *  Procedure:
 *	list_res_name - displays packet contents to stderr
 *
 ***********************************************************************/
void list_res_name(unsigned long *body)
{
  fprintf(stderr,"class resource name\n");
  fprintf(stderr,"\t ID %lx\n",body[0]);
  fprintf(stderr,"\t frame ID %lx\n",body[1]);
  fprintf(stderr,"\t fvwm ptr %lx\n",body[2]);
  fprintf(stderr,"\t resource name %s\n",(char *)(&body[3]));
}

/***********************************************************************
 *
 *  Procedure:
 *	list_end - displays packet contents to stderr
 *
 ***********************************************************************/
void list_end(void)
{
  fprintf(stderr,"Send_WindowList End\n");
}
